### PR TITLE
Send customized memo to QBO

### PR DESCRIPTION
**Pull request by Sravanksk**

### CODE REVIEW

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  getTitle(name: string) {
+    return name.replace(/_/g, ' ');
+  }
+
+  showPreview(memoStructure) {
+    const time = Date.now();
+    const today = new Date(time);
+    memoStructure = memoStructure + '';
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:36]: Can we name it preview or previewText?

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    memoStructure = memoStructure.replace('employee_email', ' john.doe@acme.com ');
+    memoStructure = memoStructure.replace('category', ' Meals and Entertainment ');
+    memoStructure = memoStructure.replace('purpose', ' Client Meeting ');
+    memoStructure = memoStructure.replace('merchant', ' Pizza Hut ');
+    memoStructure = memoStructure.replace('report_number', ' C/2021/12/R/1 ');
+    memoStructure = memoStructure.replace('spent_on', today.toLocaleDateString());
+    memoStructure = memoStructure.replace(/,/g, ' - ');
+    memoStructure = memoStructure.replace(/_/g, ' ');
+    return memoStructure;
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:45]: Something like this should do for dynamic rendering as well. Pliss fix the var names.
```
arr=['email', 'category']
a={
    email: 'john@doe',
    category: 'Food'
}
text=''
arr.forEach((e, index) => {
    if (e in a) {
        text = text + a[e];
        if (index+1 !== arr.length) {
            text = text + ' - '
        }
    }

})

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  getMemoStructureSettings() {
+    const that = this;
+    that.isLoading = true;
+    that.settingsService.getGeneralSettings(this.workspaceId).subscribe(generalSettings => {
+      that.generalSettings = generalSettings;
+      that.showPreview(generalSettings.memo_structure);
+      that.isLoading = false;
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:54]: place this after form creation

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+      that.form = that.formBuilder.group({
+        memoStructure: [that.generalSettings.memo_structure ? that.generalSettings.memo_structure : this.defaultMemoFields]
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:57]: in which case `that.generalSettings.memo_structure` won't be available?

File: src/app/qbo/settings/configuration/configuration.component.html [hunk 5]
```diff
@@ -12,6 +12,9 @@
     <a mat-tab-link (click)="changeState('EXPENSE_FIELDS')" *ngIf="showExpenseFields()" [active]="state==='EXPENSE_FIELDS'">
       Expense Fields
     </a>
+    <a mat-tab-link (click)="changeState('MEMO_STRUCTURE')" *ngIf="showExpenseFields()" [active]="state==='MEMO_STRUCTURE'">
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/configuration.component.html:15]: Modify description should not depend on `showExpenseFields()` imho

Comment (ashwin1111) [reply src/app/qbo/settings/configuration/configuration.component.html:15]: It should just check for workspace_general_settings to be existed

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      });
+
+    }, () => {
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:60]: When will this case occur? Ideally we should have workspace general settings at this point right?

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    });
+  }
+
+
+  save() {
+    const that = this;
+    that.isLoading = true;
+
+    const selectedMemoFields = that.defaultMemoFields.filter(i => that.form.value.memoStructure.indexOf(i) !== -1);
+    const memoStructure = selectedMemoFields ? selectedMemoFields : that.defaultMemoFields;
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:74]: If the user doesn't select any fields, are we selecting all options by default?

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      });
+    });
+  }
+
+
+  save() {
+    const that = this;
+    that.isLoading = true;
+
+    const selectedMemoFields = that.defaultMemoFields.filter(i => that.form.value.memoStructure.indexOf(i) !== -1);
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:73]: name i better. Can we `memoOption` or something better than this.

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    const memoStructure = selectedMemoFields ? selectedMemoFields : that.defaultMemoFields;
+
+    const generalSettingsPayload: GeneralSetting = {
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:76]: need not be of type `GeneralSetting`, can just pass `string[]` to the service

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      memo_structure: memoStructure
+    };
+
+    that.settingsService.patchGeneralSettings(this.workspaceId, generalSettingsPayload).subscribe((response) => {
+      that.snackBar.open('Custom Memo saved successfully');
+      that.getMemoStructureSettings();
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:82]: Can we use the response of the call to render the form? Can reduce a get call by this way.

**Review by ashwin1111 — CHANGES_REQUESTED** [review]

File: src/app/qbo/settings/configuration/general-configuration/general-configuration.component.ts [hunk 11]
```diff
@@ -209,6 +211,7 @@ export class GeneralConfigurationComponent implements OnInit {
         changeAccountingPeriod: [false],
         importTaxCodes: [null],
         chartOfAccounts: [['Expense']],
+        customizedMemo: [null],
```
[the remark this answers is not in the document]
Comment (Sravanksk) [reply src/app/qbo/settings/configuration/general-configuration/general-configuration.component.ts:214]: removed

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.html [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+<div *ngIf="!isLoading" fxLayout="column" fxLayoutAlign="start stretch">
+  <div class="page-info">
+    <div class="page-info--description">
+      Customize the data set you would like to pass to the description field of QBO while exporting expenses from Fyle.
+    </div>
+  </div>
+  <div class="page-info">
+    <div class="memo--label">
+      Select and reorder the fields
```
[the remark this answers is not in the document]
Comment (Sravanksk) [reply src/app/qbo/settings/configuration/memo-structure/memo-structure.component.html:13]: fixed

File: src/app/qbo/settings/configuration/general-configuration/general-configuration.component.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (ashwin1111) [reply src/app/qbo/settings/configuration/general-configuration/general-configuration.component.ts:212]: Why is this needed?

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts (hunk 6 above)
Comment (ashwin1111) [reply src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:72]: Can you answer this plis

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts (hunk 7 above)
Comment (ashwin1111) [reply src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:86]: Can you pliss answer this

File: src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    const selectedMemoFields = that.defaultMemoFields.filter(memoOption => that.form.value.memoStructure.indexOf(memoOption) !== -1);
+    const memoStructure = selectedMemoFields ? selectedMemoFields : that.defaultMemoFields;
+    that.generalSettings.memo_structure = memoStructure;
+  }
+
+  getTitle(name: string) {
+    return name.replace(/_/g, ' ');
+  }
+
+  showPreview(previewText) {
```
Comment (ashwin1111) [inline src/app/qbo/settings/configuration/memo-structure/memo-structure.component.ts:37]: name `previewText` better and add type too

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**